### DISCUSSION TITLE
A value of type enum can be serialized, as long as the enum is zero-based (has a value corresponding to 0)

### DISCUSSION THREAD

**Discussion by Carpenter360** [MarimerLLC/csla#4345] [Questions] [answered]
Hi @rockfordlhotka,

I came across this quote from ebook **Using CSLA 4 Creating Business Objects**, Revision: 1.1
Table 8. Types directly serialized by MobileFormatter, page 32

"A value of type enum can be serialized, as long as the enum is zero-based (has a value corresponding to 0)"

So I have scaffold a quick FACT to force this error but my test completed successfully.
Is this a book errata, or is my test wrong, or has this restriction been lifted over the years?

Thank you in advance,

Sylvain

---

namespace Customer
{
    [Serializable]
    public class AssessmentEdit : BusinessBase<AssessmentEdit>
    {
        public static readonly PropertyInfo<long> IdProperty = RegisterProperty<long>(c => c.Id);
        public long Id
        {
            get => GetProperty(IdProperty);
            set => SetProperty(IdProperty, value);
        }

        public static readonly PropertyInfo<ExpenseEligibility> EligibilityProperty = RegisterProperty<ExpenseEligibility>(c => c.Eligibility);
        public ExpenseEligibility Eligibility
        {
            get => GetProperty(EligibilityProperty);
            set => SetProperty(EligibilityProperty, value);
        }

        [Create]
        [RunLocal]
        private void Create()
        {
            // Run local logic for creating a new AssessmentEdit object
            using (BypassPropertyChecks)
            {
                Id = -1; // Indicate a new object
                Eligibility = ExpenseEligibility.Pending; // Default eligibility
            }
            BusinessRules.CheckRules();
        }
    }
}

public enum ExpenseEligibility
{
    /// <summary>Expenses Pending.</summary>
    Pending = 1,

    /// <summary>Expenses No Determination.</summary>
    NoDetermination = 2,

    /// <summary>Expenses Eligible.</summary>
    Eligible = 3,

    /// <summary>Expenses Not Eligible.</summary>
    NotEligible = 4,
}

[Fact]
public void ShouldSerializeAndDeserializeAssessmentEdit_WhenUsingMobileFormatter()
{
    // Arrange
    var dpFactory = _fixture.ServiceProvider.GetRequiredService<IDataPortalFactory>();
    var originalAssessmentEdit = dpFactory.GetPortal<AssessmentEdit>().Create();
    originalAssessmentEdit.Eligibility = ExpenseEligibility.Eligible;

    // Act
    var serializedData = Serialize(originalAssessmentEdit);
    var deserializedAssessmentEdit = Deserialize<AssessmentEdit>(serializedData);

    // Assert
    deserializedAssessmentEdit.Eligibility.Should().Be(originalAssessmentEdit.Eligibility);
}

private byte[] Serialize(object obj)
{
    using (var stream = new MemoryStream())
    {
        var formatter = new MobileFormatter(_fixture.ApplicationContext);
        formatter.Serialize(stream, obj);
        return stream.ToArray();
    }
}

private T Deserialize<T>(byte[] data)
{
    using (var stream = new MemoryStream(data))
    {
        var formatter = new MobileFormatter(_fixture.ApplicationContext);
        return (T)formatter.Deserialize(stream);
    }
}

**Comment by StefanOssendorf** [comment, marked as answer]
I have no clue about CSLA 4 but I'm using CSLA 7+ and have some enums where this restriction does not apply. So it was probably removed over the years.

**Reply by rockfordlhotka** [reply]
A lot has changed since .NET 4 and CSLA 4, that's for sure!!

**Comment by Carpenter360** [comment]
I'm actually using CSLA 9 (R24070301)!

@rockfordlhotka please let me know if you need a reviewer for any ebook/documentation updates.

Cheers,

Sylvain
